python_ta/checkers/redundant_assignment_checker.py

```diff
@@ -0,0 +1,130 @@
+        """
+        out_facts = {}
+        cfg = ControlFlowGraph()
+        cfg.start = node.cfg_block
+        worklist = list(cfg.get_blocks_postorder())
+        worklist.reverse()
+
+        all_assigns = self._get_assigns(node)
+        for block in worklist:
+            out_facts[block] = all_assigns.copy()
+
+        while len(worklist) != 0:
+            b = worklist.pop()
+            outs = [out_facts[p.target] for p in b.successors if p.target in out_facts]
+            if outs == []:
+                in_facts = set()
+            else:
+                in_facts = set.intersection(*outs)
+            temp = self._transfer(b, in_facts, node)
+            if temp != out_facts[b]:
+                out_facts[b] = temp
+                worklist.extend([pred.source for pred in b.predecessors])
+
+    def _transfer(self, block: CFGBlock, out_facts: Set[str], scope: Union[astroid.Module, astroid.FunctionDef]) -> Set[str]:
+        gen = out_facts.copy()
+        kill = set()
+        for statement in reversed(block.statements):
+            if isinstance(statement, astroid.Expr) and isinstance(statement.value, astroid.Call) and \
+                    isinstance(statement.value.func, astroid.Name):
+                for func in (scope.locals.get(statement.value.func.name) or []):
+                    if isinstance(func, astroid.FunctionDef) and not gen.issubset(set(func.locals)):
+                        kill = kill.union(gen)
+
+            for node in statement.nodes_of_class((astroid.AssignName, astroid.DelName, astroid.Name),
+                                              astroid.FunctionDef):
+                if isinstance(node, astroid.AssignName):
+                    if node.name in gen.difference(kill):
+                        self._redundant_assignment.add(node.parent)
+                    elif node.parent in self._redundant_assignment:
+                        self._redundant_assignment.remove(node.parent)
+
+                    kill.discard(node.name)
+                    gen.add(node.name)
+                else:
+                    kill.add(node.name)
+
+        return gen.difference(kill)
+
+    def _get_assigns(self, node: Union[astroid.FunctionDef, astroid.Module]) -> Set[str]:
+        """Returns a set of all local and parameter variables that could be
+        defined in the program (either a function or module).
+
+        IF a variable 'v' is defined in a function and there is no global/nonlocal
+        statement applied to 'v' THEN 'v' is a local variable.
+
+        Note that `local variable` in the context of a module level analysis,
+        refers to global variables.
+        """
+        assigns = set()
+        kills = set()
+        for name, nodes in node.scope().locals.items():
+            if any(isinstance(elem, astroid.AssignName) for elem in nodes):
+                assigns.add(name)
+        for statement in node.nodes_of_class((astroid.Nonlocal, astroid.Global), astroid.FunctionDef):
```

Comment:
I don't see why this part is necessary. If you're doing a loop through just `node.scope().locals`, doesn't that only include local variables anyway?

Reply:
<img width="257" alt="Screen Shot 2019-11-05 at 5 52 58 PM" src="https://user-images.githubusercontent.com/18351436/68253407-da891c80-fff5-11e9-95b8-2173e411a884.png">

The `locals` attribute is not entirely accurate, that's why I have to do an extra check.

Reply:
Ah yes, we discussed this with the last checker. Given what I said above about not checking into function calls, though, you can remove this check here an include `global` and `nonlocal` variables in the search for redundant assignments.

The difference between this and the previous checker is that the only _false positives_ that would appear is if you have two assignments to the same global/nonlocal variable, but in between there's a function call that uses this variable. I don't care about checking for this situation right now.